### PR TITLE
Admin Dashboard

### PR DESCRIPTION
**Pull request by alexapostolu**
The admin dashboard. In the user's profile page, if they are an admin, they will see a button that when clicked, takes them to a sperate Admin page.

Here, they can view all the solutions, filter by utorid and timestamp. If no utorid is given, it will display solutions from any user. If no timestamp is given, it will display solutions from any time. The input box updates on every key press, so you can just type the name or the timestamp and it will apply the correct filter immediately.

In the first image, we can see the button that takes us to the Admin dashboard. In the second screenshot, we can see we are filtering by utorid and timestamp, which displays the correct results. (In both images, you can see the Admin link on the nav bar. I realized this is a bug and removed this in my most recent commit)

Screenshot 2024-04-14 101623

image

### CODE REVIEW

File: client/src/pages/Admin.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        }
+        return response.json() as Promise<SolutionData[]>;
+      })
+      .then(data => {
+			// Filter solutions based on date
+			const filteredSolutionsByDate = data.filter(solution => {
+			// Assuming solution.date is a string in format "YYYY-MM-DD"
+			const date1 = solution.createdAt.split('T')[0];
+			if (filteredUsername === "" || solution.userId === filteredUsername)
+				return filteredTimestamp === "" || date1 === filteredTimestamp;
```
Comment (arthurhohinng) [inline client/src/pages/Admin.tsx:34]: Does the timestamp filtering allow for a range?
Like show all solutions from jan 1 - jan 4

**Review by arthurhohinng — COMMENTED** [review]
Couple comments

File: client/src/pages/Admin.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      <Row className="mb-3">
+        <Col>
+          <Form onSubmit={handleFilterSubmit}>
+            <Form.Group controlId="usernameFilter">
+              <Form.Label>Filter by Username</Form.Label>
+              <Form.Control
+                type="text"
+                placeholder="Enter username"
+                value={filteredUsername}
+                onChange={handleFilterChange}
```
[the remark this answers is not in the document]
Comment (alexapostolu) [reply client/src/pages/Admin.tsx:72]: I'm not sure how to do that, because `e` isn't defined

**Comment by alexapostolu** [comment]
Fixed indentation and added a timestamp range

File: client/src/pages/Admin.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (erensunerr) [reply client/src/pages/Admin.tsx:72]: `onChange = { (e) => setFilteredUsername(e.target.value) }`

File: client/src/App.tsx [hunk 3]
```diff
@@ -1,14 +1,17 @@
...
-import Challenges from "./pages/Challenges.tsx";
-import Challenge from "./pages/Challenge.tsx";
-import Solutions from "./pages/Solutions.tsx";
-import PostSolution from "./pages/PostSolution.tsx";
-import Solution from "./pages/Solution.tsx";
-import Profile from "./pages/Profile.tsx"
-import Editor from "./pages/Editor.tsx";
+//import { useState, useEffect } from 'react';
+import { createBrowserRouter, RouterProvider } from 'react-router-dom';
+//import NavigationBar from './components/NavigationBar.tsx'; // Import the NavigationBar component
```
Comment (arthurhohinng) [inline client/src/App.tsx:3]: nit: whats with the commented lines here

File: client/src/App.tsx [hunk 4]
```diff
@@ -25,6 +28,10 @@ const NAV_CONFIG = {
       name: "Solutions",
       href: "/solutions",
     },
+    // {
+    //   name: "Admin",
+    //   href: "/admin",
+    // }
```
Comment (arthurhohinng) [inline client/src/App.tsx:34]: Same here

File: client/src/App.tsx [hunk 5]
```diff
@@ -72,13 +79,44 @@ const router = createBrowserRouter([
...
+//       });
+//   }, []);
+
+//   return (
+//     <>
+//       <RouterProvider router={router} />
+//       <NavigationBar userRole={userRole} />
+//     </>
+//   );
+// }
```
Comment (arthurhohinng) [inline client/src/App.tsx:119]: and here

**Review by arthurhohinng — COMMENTED** [review]
Comment abt the commented code

**Review by arthurhohinng — APPROVED** [review]

**Pull request merged**